File: Algo/Leetcode/041FirstMissingPositive.cpp
//1 brute force
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        set<int> s;
        for(auto n:nums){
            if(n>0) s.insert(n);
        }

        int i = 1;
        while(true){
            if(s.find(i)==s.end()) return i;
            i++;
        }
        return -1;
    }
};

//2. Use nums as a hash table. Consider those negative value.
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for(int i=0; i<n; i++)
            if(nums[i] <= 0) nums[i] = n+1;

        for(int i=0; i<n; i++)
            if(abs(nums[i]) < n+1) nums[abs(nums[i])-1] = -abs(nums[abs(nums[i])-1]);

        for(int i=0; i<n; i++)
            if(nums[i] > 0) return i+1;

        return n+1;
    }
};

File: Algo/Leetcode/041FirstMissingPositive.py
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        for i in range(n):
            if nums[i] <= 0:
                nums[i] = n + 1

        for i, k in enumerate(nums):
            if abs(k) < n + 1:
                nums[abs(k) - 1] = - abs(nums[abs(k) - 1])

        for i, k in enumerate(nums):
            if k > 0:
                return i + 1

        return n + 1